README.md
# Air setting groups

## Registering your custom setting groups

To begin using your custom setting groups, you need to create a custom setting group post. You can do this in the admin panel CPT that the plugin creates automatically.

After the setting group is created, it needs to be registered in `functions.php`.

Like this.
```php
'custom_settings' => [
    'your-custom-setting' => [
      'id' => '{YOUR POST ID}',
      'title' => 'Your custom setting',
      'block_editor' => true,
    ],
  ],
```

If you set `block-editor` to `true`, your custom setting group will support the gutenberg block editor. If you dont want it enabled you can remove `block-editor` from the array or just change its value to something that doesnt evaluate to `true`.

Now that you have registered your custom setting group, you can find it in ACF field group location rules with the title that you gave to it.

## Functions and hooks

`get_plugin_version()` Returns the version of the plugin.

`get_prefix()` Returns the prefix of the CPT. If given argument `true`, returns prefix with hyphens instead of underscores.

`get_custom_setting_config()` Returns all of the registered custom setting groups in `functions.php`

`get_custom_setting( 'key of your ACF field', 'key of your custom setting group' )` Gets you the value of a specific field of a specific setting group.

`add_filters( 'air_setting_groups_prefix', $prefix)` Filter will let you change the prefix for the plugin. This is used as the slug of setting group custom post type.



CHANGELOG.md
### 1.0.2: 2023-12-01

* Open Changelog
* Change menu_name label to more understandable in context
* Add .editorconfig 


acf-location-rule.php
<?php
/**
 * @Author:		Elias Kautto
 * @Date:   		2022-01-26 13:11:12
 * @Last Modified by:   Timi Wahalahti
 * @Last Modified time: 2023-03-16 12:54:44
 *
 * @package air-setting-groups
 */

namespace Air_Setting_Groups;

function init_settings_page_acf_location_rule() {
  $group_post_ids = get_custom_setting_config();

  if ( ! function_exists( 'acf_register_location_type' ) || empty( $group_post_ids ) ) {
    return;
  }

  acf_register_location_type( __NAMESPACE__ . '\ACF_Location_Custom_Settings' );
} // end init_settings_page_acf_location_rule

class ACF_Location_Custom_Settings extends \ACF_Location {
  public function initialize() {
    $this->name = get_prefix();
    $this->label = __( 'Asetusryhmä', get_prefix() ); // phpcs:ignore
    $this->category = 'page';
  }

  /**
   * Check if current post key exists in custom setting group lists.
   */
  public function match( $rule, $screen, $field_group ) {
    $post_id = isset( $screen['post_id'] ) ? intval( $screen['post_id'] ) : null;

    if ( ! $post_id || get_prefix( true ) !== get_post_type( $post_id ) ) {
      return false;
    }

    $setting_pages = get_all_custom_setting_pages_for_key( $rule['value'] );
    // Compare setting group key to the original setting group post
    if ( in_array( $post_id, $setting_pages ) ) { // phpcs:ignore
      return true;
    }

    return false;
  } // end match

  public static function get_operators( $rule ) {
    return array(
      '==' => __( 'is equal to', 'acf' ), // phpcs:ignore
    );
  } // end get_operators

  /**
   * Get the titles of custom settings for ACF field groups.
   */
  public function get_values( $rule ) {
    $settings = get_custom_setting_config();
    $choices = [];

    foreach ( $settings as $key => $data ) {
      $choices[ $key ] = $data['title'];
    }

    return $choices;
  } // end get_values
}


air-setting-groups.php
<?php
/**
 * Plugin Name: Air setting groups
 * Description: Provides a way to store custom settings in CPT and receive those with field id and setting group key.
 * Plugin URI: https://dude.fi
 * Author: Digitoimisto Dude Oy
 * Author URI: https://dude.fi
 * Version: 1.0.2
 * License: GPL2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 * Network: false
 * Text Domain: air_setting_groups
 *
 * @package air-setting-groups
 */

namespace Air_Setting_Groups;

defined( 'ABSPATH' ) || exit;

function get_plugin_version() {
  return 101;
} //end get_plugin_version

function get_prefix( $hyphens = false ) {
  $prefix = 'air_setting_groups';
  $prefix = apply_filters( 'air_setting_groups_prefix', $prefix );

  if ( $hyphens ) {
    $prefix = str_replace( '_', '-', $prefix );
  }

  return $prefix;
} //end get_prefix

/**
 * Gets custom settings array from theme settings.
 */
function get_custom_setting_config( $post_ids = [] ) {
  if ( ! defined( 'THEME_SETTINGS' ) ) {
    return $post_ids;
  }

  if ( ! isset( THEME_SETTINGS['custom_settings'] ) ) {
    return $post_ids;
  }

  return wp_parse_args( THEME_SETTINGS['custom_settings'], $post_ids );
} // end get_custom_setting_config

/**
 * Get custom setting pages for a specific post id.
 */
function get_all_custom_setting_pages_for_key( $key ) {
  $setting_group_posts = get_custom_setting_config();
  $original_setting_post = null;

  foreach ( $setting_group_posts as $group => $data ) {
    if ( $key === $group ) {
      $original_setting_post = $data['id'];
    }
  };

  if ( ! function_exists( 'pll_get_post_translations' ) ) {
    return [
      $original_setting_post,
    ];
  }

  if ( pll_is_translated_post_type( get_prefix( true ) ) ) {
    $translations = pll_get_post_translations( $original_setting_post );
    return $translations;
  }

  return [ $original_setting_post ];
} // end get_all_custom_setting_pages_for_key

/**
 * Add setting group CPT to polylang allow multilingual settings.
 */
add_filter( 'pll_get_post_types', function( $post_types, $is_settings ) {
  $post_types[ get_prefix( true ) ] = get_prefix( true );
  return $post_types;
}, 10, 2 );

/**
 * Register CPT.
 */
include plugin_dir_path( __FILE__ ) . '/cpt-settings.php';
add_action( 'init', __NAMESPACE__ . '\register_cpt' );

/**
 * ACF - Adds custom location rule to ACF
 * Running in plugins_loaded action to make sure that ACF is activated.
 */
add_action( 'plugins_loaded', function() {
  if ( ! class_exists( 'ACF_Location' ) ) {
    return;
  }

  include plugin_dir_path( __FILE__ ) . '/acf-location-rule.php';
  add_action( 'acf/init', __NAMESPACE__ . '\init_settings_page_acf_location_rule' );
} );

/**
 * Block editor
 */
include plugin_dir_path( __FILE__ ) . '/setting-group-block-editor.php';
add_action( 'admin_init', __NAMESPACE__ . '\air_setting_groups_editor_support_for_setting_group_post', 99, 1 );

/**
 * Get custom setting function
 */
include plugin_dir_path( __FILE__ ) . '/get-setting.php';

/**
 * Load translations
 */
add_action( 'plugins_loaded', function() {
  $plugin_path = '/air-setting-groups/languages/';
  load_plugin_textdomain( get_prefix(), false, $plugin_path );
} );


setting-group-block-editor.php
<?php
/**
 * @Author:		Elias Kautto
 * @Date:   		2022-01-26 15:31:41
 * @Last Modified by:   Elias Kautto
 * @Last Modified time: 2022-01-28 12:13:14
 *
 * @package air-setting-groups
 */

namespace Air_Setting_Groups;

/**
 * Remove editor from post type supports if its not enabled
 */
function air_setting_groups_editor_support_for_setting_group_post() {
  // Try find out which post id we are loading in admin
  $post_id = false;
  if ( isset( $_GET['post'] ) ) { // phpcs:ignore
    $post_id = intval( $_GET['post'] ); // phpcs:ignore
  } elseif ( isset( $_POST['post_ID'] ) ) { // phpcs:ignore
    $post_id = intval( $_POST['post_ID'] ); // phpcs:ignore
  } elseif ( isset( $_POST['post_id'] ) ) { // phpcs:ignore
    $post_id = intval( $_POST['post_id'] ); // phpcs:ignore
  }

  if ( ! $post_id ) {
    return;
  }

  if ( ! use_block_editor_in_custom_setting_group( $post_id ) ) {
    remove_post_type_support( get_prefix( true ), 'editor' );
  }
} // end air_setting_groups_editor_support_for_setting_group_post

/**
 * Check if post has block editor enabled, return true if enabled.
 */
function use_block_editor_in_custom_setting_group( $post_id ) {
  if ( get_prefix( true ) !== get_post_type( $post_id ) ) {
    return false;
  }

  $setting_group_post_ids = get_custom_setting_config();
  $found = false;

  foreach ( $setting_group_post_ids as $key => $data ) {
    if ( false === pll_get_post( $data['id'] ) || null === pll_get_post( $data['id'] ) ) {
      continue;
    }

    $setting_group_post_ids[ $key ]['id'] = pll_get_post( $data['id'] );
  }

  foreach ( $setting_group_post_ids as $data ) {
    if ( $data['id'] !== $post_id ) {
      continue;
    }

    $found = true;
  }

  if ( false === $found ) {
    return false;
  }

  foreach ( $setting_group_post_ids as $key => $data ) {
    if ( ! array_key_exists( 'block_editor', $data ) ) {
      unset( $setting_group_post_ids[ $key ] );
    }
  }

  foreach ( $setting_group_post_ids as $key => $data ) {
    if ( $post_id === $data['id'] && true === $data['block_editor'] ) {
      return true;
    }
  }

  return false;
} // end use_block_editor_in_custom_setting_group

cpt-settings.php
<?php
/**
 * @Author:		Elias Kautto
 * @Date:   		2022-01-26 11:19:51
 * @Last Modified by:   Roni Laukkarinen
 * @Last Modified time: 2023-12-01 12:03:30
 *
 * @package air-setting-groups
 */

namespace Air_Setting_Groups;

function register_cpt() {
  $generated_labels = [
    'menu_name'          => __( 'Other content', 'air_setting_groups' ),
    'name'               => _x( 'Setting groups', 'post type general name', 'air_setting_groups' ),
    'singular_name'      => _x( 'Setting group', 'post type singular name', 'air_setting_groups' ),
    'name_admin_bar'     => _x( 'Setting groups', 'add new on admin bar', 'air_setting_groups' ),
    'add_new'            => _x( 'Add new', 'thing', 'air_setting_groups' ),
    'add_new_item'       => __( 'Add new setting group', 'air_setting_groups' ),
    'new_item'           => __( 'New setting group', 'air_setting_groups' ),
    'edit_item'          => __( 'Edit a setting group', 'air_setting_groups' ),
    'view_item'          => __( 'Show setting group', 'air_setting_groups' ),
    'all_items'          => __( 'Show all setting groups', 'air_setting_groups' ),
    'search_items'       => __( 'Search for setting groups', 'air_setting_groups' ),
    'parent_item_colon'  => __( 'Setting group parents', 'air_setting_groups' ),
    'not_found'          => __( 'Setting groups not found', 'air_setting_groups' ),
    'not_found_in_trash' => __( 'Setting groups not found in the trash', 'air_setting_groups' ),
  ];

  $args = [
    'labels'              => $generated_labels,
    'description'         => '',
    'menu_position'       => 55,
    'menu_icon'           => 'dashicons-forms',
    'public'              => false,
    'has_archive'         => false,
    'exclude_from_search' => false,
    'show_ui'             => true,
    'show_in_menu'        => true,
    'show_in_rest'        => true,
    'rewrite'             => false,
    'supports'            => [ 'title', 'revisions', 'editor' ],
    'taxonomies'          => [],
  ];

  register_post_type( get_prefix( true ), $args );
}